File: src/api/handlers/webhooks.rs
#[cfg(feature = "web")]
use axum::{Json, Router, routing::get};
use serde::{Deserialize, Serialize};

/// Webhook response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WebhookResponse {
    pub id: String,
    pub name: String,
    pub url: String,
    pub events: Vec<String>,
    pub active: bool,
    pub secret_configured: bool,
    pub last_triggered: Option<String>,
    pub failure_count: u32,
    pub created_at: String,
}

/// Create webhook request
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateWebhookRequest {
    pub name: String,
    pub url: String,
    pub events: Vec<String>,
    pub secret: Option<String>,
}

#[cfg(feature = "web")]
pub fn router() -> Router {
    Router::new()
        .route("/webhooks", get(list_webhooks).post(create_webhook))
}

#[cfg(feature = "web")]
async fn list_webhooks() -> Json<Vec<WebhookResponse>> {
    Json(vec![])
}

#[cfg(feature = "web")]
async fn create_webhook(Json(req): Json<CreateWebhookRequest>) -> Json<Option<WebhookResponse>> {
    let _ = req;
    Json(None)
}

File: web/src/api/webhooks.ts
// Copyright 2026 Zyvor AI Labs · https://zyvor.dev
// SPDX-License-Identifier: Apache-2.0

import { apiGet, apiPost, apiDelete } from './client'

/** Mirrors src/api/http_server/web/webhook_handlers.rs, backed by the real
 * crate::api::webhooks::WebhookManager -- SSRF-guarded (HTTPS + no
 * private/internal hosts), disk-persisted, and actually delivered via a
 * real HTTPS POST with retry/backoff. The secret (if set) goes in an
 * X-Zorvia-Webhook-Secret header, not an HMAC body signature. */
export type WebhookEventName =
  | 'vm.created' | 'vm.deleted' | 'vm.started' | 'vm.stopped' | 'vm.restarted' | 'vm.failed'
  | 'vm.health_changed' | 'snapshot.created' | 'snapshot.restored' | 'backup.completed'
  | 'backup.failed' | 'migration.started' | 'migration.completed' | 'alert.triggered'
  | 'alert.resolved' | string

export interface Webhook {
  id: string
  name: string
  url: string
  events: WebhookEventName[]
  enabled: boolean
  retry_count: number
  retry_delay_secs: number
  timeout_secs: number
  created_at: string
  last_triggered?: string | null
  delivery_count: number
  failure_count: number
  success_rate: number
}

export interface CreateWebhookRequest {
  name: string
  url: string
  events: WebhookEventName[]
  secret?: string
  retry_count?: number
  retry_delay_secs?: number
}

const API_BASE = '/api'

export async function listWebhooks(): Promise<Webhook[]> {
  return apiGet<Webhook[]>(`${API_BASE}/webhooks`)
}

export async function createWebhook(req: CreateWebhookRequest): Promise<Webhook> {
  return apiPost<Webhook>(`${API_BASE}/webhooks`, req)
}

export async function deleteWebhook(id: string): Promise<void> {
  return apiDelete(`${API_BASE}/webhooks/${encodeURIComponent(id)}`)
}

export async function testWebhook(id: string): Promise<{ delivered: boolean; success: boolean }> {
  return apiPost<{ delivered: boolean; success: boolean }>(`${API_BASE}/webhooks/${encodeURIComponent(id)}/test`)
}
